src/rgw/rgw_notify.cc

```diff
@@ -2,17 +2,530 @@
+        return ret;
+      }
+      list_of_queues_object_created = true;
+    }
+    return 0;
+  }
+
+  // read the list of queues from the queue list object
+  int read_queue_list(queues_t& queues) {
+    bufferlist bl;
+    constexpr auto chunk_size = 1024U;
+    auto start_offset = 0U;
+
+    int ret;
+    do {
+      bufferlist chunk_bl;
+      // TODO: add yield ?
+      ret = rados_ioctx.read(Q_LIST_OBJECT_NAME, chunk_bl, chunk_size, start_offset);
+      if (ret < 0) {
+        ldout(cct, 1) << "ERROR: failed to read queue list. error: " << ret << dendl;
+        return ret;
+      }
+      start_offset += ret;
+      bl.claim_append(chunk_bl);
+    } while (ret > 0);
+
+    if (bl.length() == 0) {
+      // nothing in the list
+      return 0;
+    }
+
+    auto iter = bl.cbegin();
+    try {
+      decode(queues, iter);
+    } catch (buffer::error& err) {
+      ldout(cct, 1) << "ERROR: failed to decode queue list. error: " << err.what() << dendl;
+      return -EINVAL;
+    }
+
+    return 0;
+  }
+
+  // set m1 to be the minimum between m1 and m2
+  static int set_min_marker(std::string& m1, const std::string m2) {
+    cls_queue_marker mr1;
+    cls_queue_marker mr2;
+    if (mr1.from_str(m1.c_str()) < 0 || mr2.from_str(m2.c_str()) < 0) {
+      return -EINVAL;
+    }
+    if (mr2.gen <= mr1.gen && mr2.offset < mr1.offset) {
+      m1 = m2;
+    }
+    return 0;
+  }
+
+   class tokens_waiter {
+     const boost::posix_time::time_duration infinite_duration = boost::posix_time::hours(1000);
+     size_t pending_tokens = 0;
+     boost::asio::deadline_timer timer;
+ 
+     struct token {
+       tokens_waiter& waiter;
+       token(tokens_waiter& _waiter) : waiter(_waiter){
+         ++waiter.pending_tokens;
+       }   
+  
+       ~token() {
+         --waiter.pending_tokens;
+         if (waiter.pending_tokens == 0) {
+           waiter.timer.cancel();
+         }   
+       }   
+     };
+  
+   public:
+
+     tokens_waiter(boost::asio::io_context& io_context) : timer(io_context) {}  
+ 
+     void async_wait(spawn::yield_context yield) { 
+       timer.expires_from_now(infinite_duration);
+       boost::system::error_code ec; 
+       timer.async_wait(yield[ec]);
+       // TODO: make sure that the timer was cancelled and not rexpired
+     }   
+ 
+     token make_token() {    
+       return token(*this);
+     }   
+  };
+
+
+  // processing of a specific queue
+  // TODO: use string_view for queue_name
+  void process_queue(spawn::yield_context yield, std::string queue_name) {
+    const auto max_elements = 1024;
+    const std::string start_marker;
+    bool truncated = false;
+    std::string end_marker;
+    std::vector<cls_queue_entry> entries;
+    auto is_idle = true;
+    auto total_entries = 0U;
+    auto pending_work = 0U;
+    tokens_waiter waiter(io_context);
+
+    const auto ret = cls_2pc_queue_list_entries(rados_ioctx, queue_name, start_marker, max_elements, entries, &truncated, end_marker);
```

Comment:
agree, but this will require a different cls_2pc api. would prefer to do that in a separate PR.
also note that rados io is going to be faster than the notification related io

Reply:
done here: https://github.com/ceph/ceph/pull/35020